DOP-5124: reconstruct table data using tanstack tables

seungpark opened the pull request:
### Stories/Links:

DOP-5124

### Current Behavior:

[test page for tables on main branch](https://docs-mongodb-org-stg.s3.us-east-2.amazonaws.com/sandbox/cloud-docs/seung.park/main/test-tables/index.html) - includes all of Atlas docs as well as the test page (sourced [here](https://github.com/rayangler/cloud-docs/blob/sandbox/source/test-tables.txt))

### Staging Links:
[test page for tables on branch](https://docs-mongodb-org-stg.s3.us-east-2.amazonaws.com/DOP-5124-test-tables/cloud-docs/seung.park/DOP-5124/test-tables/index.html)



### Notes:
- This PR does not change the way tables are rendered, or how they should be used in ReST. This updates the logic for constructing table data on the front end, so that we can pass processed column/row data, enabling us to use farther features as [expanded content ](https://github.com/mongodb/leafygreen-ui/tree/main/packages/table#data--renderexpandedcontent) and/or nested tables.

### README updates

- - [ ] This PR introduces changes that should be reflected in the README, and I have made those updates.
- - [X] This PR does not introduce changes that should be reflected in the README

biancalaube reviewed (commented):
I noticed that in the staging link the `Nested table with header rows` had some weird spacing. I think both the columns in the staged link are the same size, but in the main link the `service` column has a smaller width. Is this expected or do we want it to look as it does originally? 
new link ----
![Screenshot 2024-12-03 at 4 27 56 PM](https://github.com/user-attachments/assets/43ea6e72-87a3-4489-a49f-77b4e8244d9b)
old link ----- 
![Screenshot 2024-12-03 at 4 30 45 PM](https://github.com/user-attachments/assets/a5767d92-638d-4538-a79e-0884966388a9)

seungpark commented:
> I noticed that in the staging link the `Nested table with header rows` had some weird spacing. I think both the columns in the staged link are the same size, but in the main link the `service` column has a smaller width. Is this expected or do we want it to look as it does originally? new link ---- ![Screenshot 2024-12-03 at 4 27 56 PM](https://private-user-images.githubusercontent.com/76073842/392144230-43ea6e72-87a3-4489-a49f-77b4e8244d9b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMyNjM0MDgsIm5iZiI6MTczMzI2MzEwOCwicGF0aCI6Ii83NjA3Mzg0Mi8zOTIxNDQyMzAtNDNlYTZlNzItODdhMy00NDg5LWE0OWYtNzdiNGU4MjQ0ZDliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjAzVDIxNTgyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEyOGU3MGM0NjQ1MmVhNTE5MWJkNGIyYTQ1MDgzODc3NWE3N2RiNDUyOWE5NzA2ZDNlYzUzMjljOTdhYTE2MTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qj1EVC5QBArwrfkT0tMeKbLAozRrokFuPJfzRo1vNyM) old link ----- ![Screenshot 2024-12-03 at 4 30 45 PM](https://private-user-images.githubusercontent.com/76073842/392144441-a5767d92-638d-4538-a79e-0884966388a9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMyNjM0MDgsIm5iZiI6MTczMzI2MzEwOCwicGF0aCI6Ii83NjA3Mzg0Mi8zOTIxNDQ0NDEtYTU3NjdkOTItNjM4ZC00NTM4LWE3OWUtMDg4NDk2NjM4OGE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjAzVDIxNTgyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiMzU5N2ZjMjdjODcyYzk5NjE0NTNjNDJkZWI0MTJlN2U3YjRlNjI1MDZkOTRjOTJkYjMxMjE1Y2FlMTIyN2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ARbyWpuOQIZpvhdASoYqpWp5av6kE5VLTxKLgiuYmNM)

addressed! [removed the styling](https://github.com/mongodb/snooty/pull/1316/commits/296ab80c1e7cadb09d707e6196e03f9e555438ef) - this was actually put in place to fix an existing issue, but that was resolved by getting the latest commit from cloud docs (seems like it was a source ReST issue)

updated the staging links above

seungpark commented on src/components/ListTable.js:
```diff
@@ -159,43 +177,69 @@ const includesNestedTable = (rows) => {
 
   // Check if :header-rows: 0 is specified or :header-rows: is omitted
-  const headerRows = headerRowCount > 0 ? children[0].children[0].children.slice(0, headerRowCount) : [];
+  const [headerRows] = useState(() =>
```
i actually initially had used useMemo hooks with dependencies that should not change (derived from nodeData) but noticed this was pretty taxing and causing page freezes. i opted to switch to a state that only initializes and does not update in [this commit](https://github.com/mongodb/snooty/pull/1316/commits/674f0901f5db93a612b936454969d2522b8195a0)

rayangler replied:
Interesting. I wonder if the bad performance was due to `bodyRows` not being memoized? Seems like we're slicing an array, which results in a new array that was potentially causing dependencies to always change?

Would something like:

```js
const bodyRows = useMemo(() => children[0].children.slice(headerRowCount), [children, headerRowCount]);
```

work?

seungpark replied:
good call! done and updated the above link

seungpark commented on src/components/ListTable.js:
```diff
@@ -159,43 +177,69 @@ const includesNestedTable = (rows) => {
+  if (!headerRow?.children) {
+    // generate columns from bodyRows
+    const flattenedRows = bodyRows.map((bodyRow) => bodyRow.children[0].children);
+    const maxColumns = Math.max(...flattenedRows.map((row) => row.length));
```
i recall an index error where a row's element couldnt find a column reference in [this line](https://github.com/mongodb/snooty/pull/1316/files#diff-c0635850a29537b5e9333b6e82a9306b70f036a6c42f9c129a903418a00ac4ebR216) so i added this safeguard.

good call on the column count! will use this output of this function to calculate that and consolidate the two

seungpark replied:
merged the logic of column count!

seungpark commented on src/components/ListTable.js:
```diff
@@ -159,43 +177,69 @@ const includesNestedTable = (rows) => {
 const ListTable = ({ nodeData: { children, options }, ...rest }) => {
   const ancestors = useAncestorComponentContext();
-  const { theme: siteTheme } = useDarkMode();
+  // TODO: header row count should not be more than 1
+  // need a warning in parser
```
done!
